### PR TITLE
bpo-45045: Optimize mapping patterns of structural pattern matching

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue45045

### CODE REVIEW

**Comment by corona10** [comment]
`````

+---------------+--------+----------------------+
| Benchmark     | base   | opt                  |
+===============+========+======================+
| bench pattern | 482 ns | 417 ns: 1.15x faster |
+---------------+--------+----------------------+
`````

File: Python/ceval.c [hunk 1]
```diff
@@ -859,7 +859,7 @@ match_keys(PyThreadState *tstate, PyObject *map, PyObject *keys)
     if (dummy == NULL) {
         goto fail;
     }
-    values = PyList_New(0);
+    values = PyTuple_New(nkeys);
```
Comment (corona10) [inline Python/ceval.c:868]: The size of the tuple is predictable.

File: Python/ceval.c [hunk 2]
```diff
@@ -873,7 +873,8 @@ match_keys(PyThreadState *tstate, PyObject *map, PyObject *keys)
             }
             goto fail;
         }
-        PyObject *value = PyObject_CallFunctionObjArgs(get, key, dummy, NULL);
+        PyObject *args[] = { key, dummy };
+        PyObject *value = PyObject_Vectorcall(get, args, 2, NULL);
```
Comment (corona10) [inline Python/ceval.c]: Just replacing PyObject_CallFunctionObjArgs shows a 2% performance enhancement on the micro benchmark.

**Comment by Fidget-Spinner** [comment]
The changes LGTM. Tested locally:
```
python -m test test_patma -R 3:3
0:00:00 Run tests sequentially
0:00:00 [1/1] test_patma
beginning 6 repetitions
123456
......

== Tests result: SUCCESS ==

1 test OK.
```
BTW, I was thinking if using ``_PyObject_GetMethod``  instead of ``_PyObject_GetAttrId`` will make your benchmark faster? The diff from your current is not too large:
```diff
@@ -846,7 +846,9 @@ match_keys(PyThreadState *tstate, PyObject *map, PyObject *keys)
     // - Don't cause key creation or resizing in dict subclasses like
     //   collections.defaultdict that define __missing__ (or similar).
     _Py_IDENTIFIER(get);
-    PyObject *get = _PyObject_GetAttrId(map, &PyId_get);
+    PyObject *get_name = _PyUnicode_FromId(&PyId_get); // borrowed
+    PyObject *get = NULL;
+    int meth_found = _PyObject_GetMethod(map, get_name, &get);
     if (get == NULL) {
         goto fail;
     }
@@ -873,8 +875,14 @@ match_keys(PyThreadState *tstate, PyObject *map, PyObject *keys)
             }
             goto fail;
         }
-        PyObject *args[] = { key, dummy };
-        PyObject *value = PyObject_Vectorcall(get, args, 2, NULL);
+        PyObject *args[] = { map, key, dummy };
+        PyObject *value = NULL;
+        if (meth_found) {
+            value = PyObject_Vectorcall(get, args, 3, NULL);
+        }
+        else {
+            value = PyObject_Vectorcall(get, &args[1], 2, NULL);
+        }
         if (value == NULL) {
             goto fail;
         }
```

**Comment by corona10** [comment]
@Fidget-Spinner 
Yeah it's better!

``````

➜  cpython git:(bpo-45045) ✗ ./python.exe -m pyperf compare_to --table base.json suggestion.json
+---------------+--------+----------------------+
| Benchmark     | base   | suggestion           |
+===============+========+======================+
| bench pattern | 482 ns | 373 ns: 1.29x faster |
+---------------+--------+----------------------+
➜  cpython git:(bpo-45045) ✗ ./python.exe -m pyperf compare_to --table opt.json suggestion.json
+---------------+--------+----------------------+
| Benchmark     | opt    | suggestion           |
+===============+========+======================+
| bench pattern | 417 ns | 373 ns: 1.12x faster |
+---------------+--------+----------------------+
``````

**Comment by corona10** [comment]
With new commit
````
0:00:00 load avg: 5.05 Run tests sequentially
0:00:00 load avg: 5.05 [1/1] test_patma
beginning 6 repetitions
123456
......

== Tests result: SUCCESS ==

1 test OK.

Total duration: 1.1 sec
Tests result: SUCCESS
````

**Review by Fidget-Spinner — APPROVED** [review]
LGTM. Thanks!

**Comment by corona10** [comment]
@Fidget-Spinner Thanks for the review.

Here is the final benchmark with optimization build with thin LTO :)

+---------------+---------------+----------------------+
| Benchmark     | thin_lto_base | thin_lto_opt         |
+===============+===============+======================+
| bench pattern | 357 ns        | 287 ns: 1.24x faster |
+---------------+---------------+----------------------+

**Pull request merged**